File: src/test/java/io/github/aalsanie/codes/RegistryMappingExceptionTest.java
package io.github.aalsanie.codes;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import java.util.NoSuchElementException;
import org.junit.jupiter.api.Test;

class RegistryMappingExceptionTest {
    @Test
    void registryIsImmutableAndKeyedByCode() {
        OutcomeDefinition custom = OutcomeDefinition.custom(
            "com.example.orders", "ORDER_REJECTED", OutcomeState.FAILED, "Rejected."
        );
        OutcomeRegistry empty = OutcomeRegistry.empty();
        OutcomeRegistry registry = empty.with(custom);

        assertEquals(0, empty.getSize());
        assertEquals(1, registry.getSize());
        assertTrue(registry.contains(custom.getCode()));
        assertSame(custom, registry.find(custom.getCode()));
        assertSame(custom, registry.find(custom.getCode().getValue()));
        assertNull(registry.find("invalid"));
        assertSame(custom, registry.require(custom.getCode()));
        assertEquals(List.of(custom), registry.definitions());
        assertThrows(UnsupportedOperationException.class, () -> registry.definitions().add(StandardOutcomes.OK));
        assertThrows(IllegalArgumentException.class, () -> registry.with(custom));
        assertThrows(NoSuchElementException.class, () -> registry.require(StandardOutcomes.NOT_FOUND.getCode()));
    }

    @Test
    void registryFactoriesComposeDefinitions() {
        OutcomeDefinition first = OutcomeDefinition.custom("com.example", "A", OutcomeState.FAILED, "A.");
        OutcomeDefinition second = OutcomeDefinition.custom("com.example", "B", OutcomeState.FAILED, "B.");
        assertEquals(2, OutcomeRegistry.of(first, second).getSize());
        assertEquals(StandardOutcomes.all.size(), OutcomeRegistry.standard().getSize());
        assertThrows(NullPointerException.class, () -> OutcomeRegistry.of((OutcomeDefinition[]) null));
        assertThrows(NullPointerException.class, () -> OutcomeRegistry.empty().with(null));
        assertThrows(NullPointerException.class, () -> OutcomeRegistry.empty().withAll(null));
    }

    @Test
    void exceptionRequiresFailedOutcomeAndPreservesCause() {
        Outcome failure = Outcome.of(StandardOutcomes.INTERNAL);
        RuntimeException cause = new RuntimeException("cause");
        OutcomeException exception = new OutcomeException(failure, cause);
        assertSame(failure, exception.getOutcome());
        assertSame(cause, exception.getCause());
        assertEquals(failure.getMessage(), exception.getMessage());
        assertSame(failure, OutcomeExceptions.toException(failure).getOutcome());
        assertSame(cause, OutcomeExceptions.toException(failure, cause).getCause());
        assertThrows(IllegalArgumentException.class, () -> new OutcomeException(Outcome.of(StandardOutcomes.OK)));
        assertThrows(NullPointerException.class, () -> new OutcomeException(null));
    }
}

File: src/test/kotlin/io/github/aalsanie/codes/RegistryMappingExceptionTest.kt
package io.github.aalsanie.codes

import org.junit.jupiter.api.Test

class RegistryMappingExceptionTest {
    private val custom = OutcomeDefinition.custom("com.example", "CUSTOM", OutcomeState.FAILED, "Custom failure.")

    @Test fun emptyRegistryIsEmpty() { assertEquals(0, OutcomeRegistry.empty().size) }
    @Test fun standardRegistryContainsAllStandards() { assertEquals(StandardOutcomes.all.size, OutcomeRegistry.standard().size) }

    @Test fun registryFindsByCodeAndString() {
        val registry = OutcomeRegistry.of(custom)
        assertSame(custom, registry.find(custom.code))
        assertSame(custom, registry.find(custom.code.value))
        assertNull(registry.find("invalid"))
    }

    @Test fun registryContainsCode() { assertTrue(OutcomeRegistry.of(custom).contains(custom.code)) }

    @Test fun registryRequireReturnsDefinition() { assertSame(custom, OutcomeRegistry.of(custom).require(custom.code)) }

    @Test fun registryRequireRejectsMissingCode() {
        assertFails<NoSuchElementException> { OutcomeRegistry.empty().require(custom.code) }
    }

    @Test fun registryRejectsDuplicateCode() {
        val duplicate = OutcomeDefinition.custom(custom.code, OutcomeState.FAILED, "Other message.")
        assertFails<IllegalArgumentException> { OutcomeRegistry.of(custom, duplicate) }
    }

    @Test fun registryWithIsPersistent() {
        val empty = OutcomeRegistry.empty()
        val next = empty.with(custom)
        assertEquals(0, empty.size)
        assertEquals(1, next.size)
    }

    @Test fun registryWithAllPreservesOrder() {
        val second = OutcomeDefinition.custom("com.example", "SECOND", OutcomeState.SUCCEEDED, "Second.")
        val registry = OutcomeRegistry.empty().withAll(listOf(custom, second))
        assertEquals(listOf(custom, second), registry.definitions())
    }

    @Test fun registryDefinitionsAreUnmodifiable() {
        @Suppress("UNCHECKED_CAST")
        val values = OutcomeRegistry.of(custom).definitions() as MutableList<OutcomeDefinition>
        assertFails<UnsupportedOperationException> { values.clear() }
    }

    @Test fun mappedResultExposesValue() {
        val mapped = MappingResult.mapped("x")
        assertTrue(mapped.isMapped)
        assertFalse(mapped.isUnmapped)
        assertEquals("x", mapped.orNull())
        assertEquals("mapped:x", mapped.fold({ "mapped:$it" }, { "none" }))
    }

    @Test fun unmappedResultExposesAbsence() {
        val unmapped = MappingResult.unmapped<String>()
        assertFalse(unmapped.isMapped)
        assertTrue(unmapped.isUnmapped)
        assertNull(unmapped.orNull())
        assertEquals("none", unmapped.fold({ "mapped:$it" }, { "none" }))
    }

    @Test fun mappedResultHasValueEquality() {
        assertEquals(MappingResult.mapped("x"), MappingResult.mapped("x"))
        assertEquals(MappingResult.mapped("x").hashCode(), MappingResult.mapped("x").hashCode())
        assertNotEquals(MappingResult.mapped("x"), MappingResult.mapped("y"))
        assertTrue(MappingResult.mapped("x").toString().contains("x"))
    }

    @Test fun exceptionRequiresFailedOutcome() {
        assertFails<IllegalArgumentException> { OutcomeException(Outcome.of(StandardOutcomes.OK)) }
    }

    @Test fun exceptionUsesStableMessageAndCause() {
        val cause = IllegalStateException("root")
        val outcome = Outcome.of(StandardOutcomes.INTERNAL, "sensitive internal detail")
        val exception = outcome.toException(cause)
        assertSame(outcome, exception.outcome)
        assertSame(cause, exception.cause)
        assertEquals(StandardOutcomes.INTERNAL.defaultMessage, exception.message)
        assertFalse(exception.message!!.contains("sensitive"))
    }
}
